gitws/git.py

```diff
@@ -330,35 +331,39 @@ def _clone_cache(self, url):
-                shutil.rmtree(cache)  # broken
-        # Cache Update
-        if cache.exists():
-            self._run(("fetch", "origin"), capture_output=True, cwd=cache)
-            try:
-                branch = self._run2str(("branch",), regex=_RE_BRANCH, cwd=cache)
-                self._run(("branch", f"--set-upstream-to=origin/{branch}", branch), capture_output=True, cwd=cache)
-                self._run(("merge", f"origin/{branch}"), capture_output=True, cwd=cache)
-            except subprocess.CalledProcessError:  # pragma: no cover
-                _LOGGER.warning("Cache Entry %s is broken. Removing.", key)
-                shutil.rmtree(cache)  # broken
-        # (Re-)Init Cache
-        if not cache.exists():
-            self.secho("Initializing clone-cache")
-            cache.mkdir(parents=True)
-            run(("git", "clone", "--", str(url), str(cache)))
-        _LOGGER.debug("Copy %s to  %s)", cache, self.path)
-        shutil.copytree(cache, self.path)
-        # Remove user/password credentials from cache
-        self._run(("remote", "remove", "origin"), cwd=cache)
+
+        with atomic_update_or_create_path(cache) as tmp_cache:
+            # Restore user/password credentials, repair corrupted cache
+            if tmp_cache.exists():
+                self.secho("Using clone-cache")
+                try:
+                    self._run(("remote", "add", "origin", str(url)), capture_output=True, cwd=tmp_cache)
+                    self._run(("reset", "--hard"), capture_output=True, cwd=tmp_cache)
+                    self._run(("clean", "-xdf"), capture_output=True, cwd=tmp_cache)
+                except subprocess.CalledProcessError:  # pragma: no cover
+                    _LOGGER.warning("Cache Entry %s is broken. Removing.", key)
+                    shutil.rmtree(tmp_cache)  # broken
+            # Cache Update
+            if tmp_cache.exists():
+                self._run(("fetch", "origin"), capture_output=True, cwd=tmp_cache)
+                try:
+                    branch = self._run2str(("branch",), regex=_RE_BRANCH, cwd=tmp_cache)
+                    self._run(
+                        ("branch", f"--set-upstream-to=origin/{branch}", branch), capture_output=True, cwd=tmp_cache
+                    )
+                    self._run(("merge", f"origin/{branch}"), capture_output=True, cwd=tmp_cache)
+                except subprocess.CalledProcessError:  # pragma: no cover
+                    _LOGGER.warning("Cache Entry %s is broken. Removing.", key)
+                    shutil.rmtree(tmp_cache)  # broken
+            # (Re-)Init Cache
+            if not tmp_cache.exists():
+                self.secho("Initializing clone-cache")
+                tmp_cache.mkdir(parents=True)
+                run(("git", "clone", "--", str(url), str(tmp_cache)))
+            _LOGGER.debug("Copy %s to  %s)", tmp_cache, self.path)
+            shutil.copytree(tmp_cache, self.path)
```

Comment:
Shouldn't we use another temporary directory here, just to avoid that an incomplete copy operation (process killed) never leaves garbage at `self.path`. Yes, it is my original code.

Reply:
We could, but I see little benefit here:

1. The caching is mostly interesting for CI/CD scenarios, where between pipelines, the cache is reused, but not individual workspaces. Hence, focus is to keep the cache clean. For the workspace copy:
2. That one also would "break" if a process is killed when no cache is used at all. Even worse, it could:
3. Break when caching is enabled, but a particular repo is already part of a workspace and hence, updated in place.